### ISSUE TITLE
[PATCH] APItest.xs: Unused dTHX

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13845]
Migrated from [rt.perl.org#121882](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121882) (status was 'resolved')

Searchable as RT121882$

**Comment by p5pRT** [comment]
### From @jhi

Even under threads\, this dTHX seems to be unnecessary\.

Patch attached\.

**Comment by p5pRT** [comment]
### From @jhi

<details><summary>0001-Unused-dTHX.patch</summary>

```diff
From 826f6ba0432a4aebeb240dcbcc631f07d8dc4b4c Mon Sep 17 00:00:00 2001
From: Jarkko Hietaniemi <jhi@iki.fi>
Date: Fri, 9 May 2014 10:34:09 -0400
Subject: [PATCH] Unused dTHX.

Fix for Coverity perl5 CID 49935.
---
 ext/XS-APItest/APItest.xs | 1 -
 1 file changed, 1 deletion(-)

diff --git a/ext/XS-APItest/APItest.xs b/ext/XS-APItest/APItest.xs
index a51924d..914abac 100644
--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -83,7 +83,6 @@ typedef void (freeent_function)(pTHX_ HV *, HE *);
 
 void
 test_freeent(freeent_function *f) {
-    dTHX;
     dSP;
     HV *test_hash = newHV();
     HE *victim;
-- 
1.9.2


```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, May 15\, 2014 at 08&#8203;:44&#8203;:48AM \-0700\, Jarkko Hietaniemi wrote&#8203;:
> \# New Ticket Created by  Jarkko Hietaniemi 
> \# Please include the string&#8203;:  \[perl \#121882\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121882 >
> 
> 
> Even under threads\, this dTHX seems to be unnecessary\.
> 
> Patch attached\.

> >From 826f6ba0432a4aebeb240dcbcc631f07d8dc4b4c Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi>
> Date&#8203;: Fri\, 9 May 2014 10&#8203;:34&#8203;:09 \-0400
> Subject&#8203;: \[PATCH\] Unused dTHX\.
> 
> Fix for Coverity perl5 CID 49935\.
> \-\-\-
>  ext/XS\-APItest/APItest\.xs | 1 \-
>  1 file changed\, 1 deletion\(\-\)
> 
> diff \-\-git a/ext/XS\-APItest/APItest\.xs b/ext/XS\-APItest/APItest\.xs
> index a51924d\.\.914abac 100644
> \-\-\- a/ext/XS\-APItest/APItest\.xs
> \+\+\+ b/ext/XS\-APItest/APItest\.xs
> @&#8203;@&#8203; \-83\,7 \+83\,6 @&#8203;@&#8203; typedef void \(freeent\_function\)\(pTHX\_ HV \*\, HE \*\);
>  
>  void
>  test\_freeent\(freeent\_function \*f\) \{
> \-    dTHX;
>      dSP;
>      HV \*test\_hash = newHV\(\);
>      HE \*victim;

\.\.\. which probably means that APItest\.xs ought to have a
\#define PERL\_NO\_GET\_CONTEX


\-\- 
Spock \(or Data\) is fired from his high\-ranking position for not being able
to understand the most basic nuances of about one in three sentences that
anyone says to him\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#19

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jhi

On Saturday\-201405\-17\, 9&#8203;:57\, Dave Mitchell wrote&#8203;:
> On Thu\, May 15\, 2014 at 08&#8203;:44&#8203;:48AM \-0700\, Jarkko Hietaniemi wrote&#8203;:
>> \# New Ticket Created by  Jarkko Hietaniemi
>> \# Please include the string&#8203;:  \[perl \#121882\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121882 >
>>
>>
>> Even under threads\, this dTHX seems to be unnecessary\.
>>
>> Patch attached\.
>
>> >From 826f6ba0432a4aebeb240dcbcc631f07d8dc4b4c Mon Sep 17 00&#8203;:00&#8203;:00 2001
>> From&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi>
>> Date&#8203;: Fri\, 9 May 2014 10&#8203;:34&#8203;:09 \-0400
>> Subject&#8203;: \[PATCH\] Unused dTHX\.
>>
>> Fix for Coverity perl5 CID 49935\.
>> \-\-\-
>>   ext/XS\-APItest/APItest\.xs | 1 \-
>>   1 file changed\, 1 deletion\(\-\)
>>
>> diff \-\-git a/ext/XS\-APItest/APItest\.xs b/ext/XS\-APItest/APItest\.xs
>> index a51924d\.\.914abac 100644
>> \-\-\- a/ext/XS\-APItest/APItest\.xs
>> \+\+\+ b/ext/XS\-APItest/APItest\.xs
>> @&#8203;@&#8203; \-83\,7 \+83\,6 @&#8203;@&#8203; typedef void \(freeent\_function\)\(pTHX\_ HV \*\, HE \*\);
>>
>>   void
>>   test\_freeent\(freeent\_function \*f\) \{
>> \-    dTHX;
>>       dSP;
>>       HV \*test\_hash = newHV\(\);
>>       HE \*victim;
>
> \.\.\. which probably means that APItest\.xs ought to have a
> \#define PERL\_NO\_GET\_CONTEX

T

Refreshed patch attached\.

**Comment by p5pRT** [comment]
### From @jhi

<details><summary>0001-Unused-dTHX.patch</summary>

```diff
From 5ef28470d77b4db2e6c2331437c585d10e6b5c7a Mon Sep 17 00:00:00 2001
From: Jarkko Hietaniemi <jhi@iki.fi>
Date: Fri, 9 May 2014 10:34:09 -0400
Subject: [PATCH] Unused dTHX.

Fix for Coverity perl5 CID 49935.
---
 ext/XS-APItest/APItest.xs | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ext/XS-APItest/APItest.xs b/ext/XS-APItest/APItest.xs
index a51924d..c2b02b9 100644
--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -10,6 +10,8 @@ typedef PTR_TBL_t *XS__APItest__PtrTable;
 #define croak_fail() croak("fail at " __FILE__ " line %d", __LINE__)
 #define croak_fail_ne(h, w) croak("fail %p!=%p at " __FILE__ " line %d", (h), (w), __LINE__)
 
+#define PERL_GET_NO_CONTEXT
+
 /* for my_cxt tests */
 
 #define MY_CXT_KEY "XS::APItest::_guts" XS_VERSION
@@ -83,7 +85,6 @@ typedef void (freeent_function)(pTHX_ HV *, HE *);
 
 void
 test_freeent(freeent_function *f) {
-    dTHX;
     dSP;
     HV *test_hash = newHV();
     HE *victim;
-- 
1.8.5.2 (Apple Git-48)


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat May 17 11&#8203;:23&#8203;:39 2014\, jhi wrote&#8203;:
> On Saturday\-201405\-17\, 9&#8203;:57\, Dave Mitchell wrote&#8203;:
> > \.\.\. which probably means that APItest\.xs ought to have a
> > \#define PERL\_NO\_GET\_CONTEX
> 
> T
> 
> Refreshed patch attached\.

While Dave was missing a T\, you've re\-ordered the name&#8203;:

\+\#define PERL\_GET\_NO\_CONTEXT

If you use the correct name&#8203;:

\#define PERL\_NO\_GET\_CONTEXT

and put it above the headers so it makes a difference\, APItest\.o fails to build&#8203;:

APItest\.xs&#8203;: In function ‘my\_cxt\_getsv\_interp’&#8203;:
APItest\.xs&#8203;:63&#8203;:5&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)
APItest\.xs&#8203;:63&#8203;:5&#8203;: note&#8203;: each undeclared identifier is reported only once for each function it appears in
APItest\.xs&#8203;: In function ‘test\_freeent’&#8203;:
APItest\.xs&#8203;:88&#8203;:5&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)
APItest\.xs&#8203;: In function ‘test\_op\_list\_describe\_part’&#8203;:
APItest\.xs&#8203;:445&#8203;:5&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)
APItest\.xs&#8203;: In function ‘test\_op\_list\_describe’&#8203;:
APItest\.xs&#8203;:465&#8203;:15&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)
APItest\.xs&#8203;: In function ‘test\_op\_linklist\_describe’&#8203;:
APItest\.xs&#8203;:518&#8203;:14&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)
make\[1\]&#8203;: \*\*\* \[APItest\.o\] Error 1
make\[1\]&#8203;: Leaving directory \`/home/tony/dev/perl/git/perl2/ext/XS\-APItest'
Unsuccessful make\(ext/XS\-APItest\)&#8203;: code=512 at make\_ext\.pl line 561\.
make&#8203;: \*\*\* \[lib/auto/XS/APItest/APItest\.so\] Error 2

Tony

**Comment by p5pRT** [comment]
### From @jhi

On Sunday\-201405\-18\, 21&#8203;:16\, Tony Cook via RT wrote&#8203;:
> \+\#define PERL\_GET\_NO\_CONTEXT
>
> If you use the correct name&#8203;:
>
> \#define PERL\_NO\_GET\_CONTEXT

Gaaak\.  Not the most shining of my weekends\, patch\-wise\.

> and put it above the headers so it makes a difference\, APItest\.o fails to build&#8203;:
>
> APItest\.xs&#8203;: In function ‘my\_cxt\_getsv\_interp’&#8203;:
 > APItest\.xs&#8203;:63&#8203;:5&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this function\)

\(Fails to build with usethreads\, but builds without\.\)

But\, umm\.  Well\.  What is going on\, then?  Note that the original 
one\-line removal of the dTHX \(without the PERL\_NO\_GET\_CONTEXT\) works 
fine\, either way\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun May 18 18&#8203;:40&#8203;:26 2014\, jhi wrote&#8203;:
> On Sunday\-201405\-18\, 21&#8203;:16\, Tony Cook via RT wrote&#8203;:
> > \+\#define PERL\_GET\_NO\_CONTEXT
> >
> > If you use the correct name&#8203;:
> >
> > \#define PERL\_NO\_GET\_CONTEXT
> 
> Gaaak\.  Not the most shining of my weekends\, patch\-wise\.
> 
> > and put it above the headers so it makes a difference\, APItest\.o
> > fails to build&#8203;:
> >
> > APItest\.xs&#8203;: In function ‘my\_cxt\_getsv\_interp’&#8203;:
> >  APItest\.xs&#8203;:63&#8203;:5&#8203;: error&#8203;: ‘my\_perl’ undeclared \(first use in this
> > function\)
> 
> \(Fails to build with usethreads\, but builds without\.\)
> 
> But\, umm\.  Well\.  What is going on\, then?  Note that the original
> one\-line removal of the dTHX \(without the PERL\_NO\_GET\_CONTEXT\) works
> fine\, either way\.

newHV\(\) expands to&#8203;:

&nbsp;  MUTABLE\_HV\(newSV\_type\(SVt\_PVHV\)\)

newSV\_type\(SVt\_PVHV\) expands to&#8203;:

&nbsp;  Perl\_newSV\_type\(aTHX\_ a\)

With PERL\_NO\_GET\_CONTEXT defined\, aTHX\_ expands to&#8203;:

&nbsp;  my\_perl

the local variable that dTHX defines \(or can be passed in through pTHX\)\.

Without PERL\_NO\_GET\_CONTEXT\, XSUB\.h redefines aTHX to be&#8203;:

&nbsp;  PERL\_GET\_THX

which expands to a platform specific function call to grab the context from TLS\.

So I see two possible solutions&#8203;:

a\) your original patch that removes a redundant dTHX

b\) a more complex patch that adds PERL\_NO\_GET\_CONTEXT and adds dTHX where appropriate\.

The first has the advantage that we continue to have a core maintained module which low performance requirements that will probably break if we somehow manage to break the aTHX re\-definition in XSUB\.h\.

The second may slightly\[1\] improve the performance \(and possibly reduce generated code size\) of XS&#8203;::APItest\.  But XS&#8203;::APItest is used only during testing\, and only for a small number of tests\, its performance isn't a big deal\.

Tony

\[1\] probably not noticably

**Comment by p5pRT** [comment]
### From @jhi

On Sunday\-201405\-18\, 22&#8203;:08\, Tony Cook via RT wrote&#8203;:
> a\) your original patch that removes a redundant dTHX
>
> The first has the advantage that we continue to have a core maintained module which low performance requirements that will probably break if we somehow manage to break the aTHX re\-definition in XSUB\.h\.

I think less complex is much more preferable; I don't think speeding up 
XS\-APItest is a worthwhile goal\.

**Comment by p5pRT** [comment]
### From @bulk88

On Sun May 18 19&#8203;:08&#8203;:21 2014\, tonyc wrote&#8203;:
> The first has the advantage that we continue to have a core maintained
> module which low performance requirements that will probably break if
> we somehow manage to break the aTHX re\-definition in XSUB\.h\.
> 
> The second may slightly\[1\] improve the performance \(and possibly
> reduce generated code size\) of XS&#8203;::APItest\.  But XS&#8203;::APItest is used
> only during testing\, and only for a small number of tests\, its
> performance isn't a big deal\.
> 
> Tony
> 
> \[1\] probably not noticably

I'm not reading too deeply into this\, but PERL\_NO\_GET\_CONTEXT mode and no PERL\_NO\_GET\_CONTEXT both should be tested\. The performance of XS&#8203;::APItest is irrelavent if it fails to test no PERL\_NO\_GET\_CONTEXT mode\. Not all core included XS libs do PERL\_NO\_GET\_CONTEXT yet so they wont build and they are a test for no PERL\_NO\_GET\_CONTEXT by itself\. But ideally in the future XS&#8203;::APItest would be the only no PERL\_NO\_GET\_CONTEXT code included in perl\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon May 19 12&#8203;:11&#8203;:50 2014\, bulk88 wrote&#8203;:
> On Sun May 18 19&#8203;:08&#8203;:21 2014\, tonyc wrote&#8203;:
> > The first has the advantage that we continue to have a core
> > maintained
> > module which low performance requirements that will probably break if
> > we somehow manage to break the aTHX re\-definition in XSUB\.h\.
> >
> > The second may slightly\[1\] improve the performance \(and possibly
> > reduce generated code size\) of XS&#8203;::APItest\.  But XS&#8203;::APItest is used
> > only during testing\, and only for a small number of tests\, its
> > performance isn't a big deal\.
> >
> > Tony
> >
> > \[1\] probably not noticably
> 
> I'm not reading too deeply into this\, but PERL\_NO\_GET\_CONTEXT mode and
> no PERL\_NO\_GET\_CONTEXT both should be tested\. The performance of
> XS&#8203;::APItest is irrelavent if it fails to test no PERL\_NO\_GET\_CONTEXT
> mode\. Not all core included XS libs do PERL\_NO\_GET\_CONTEXT yet so they
> wont build and they are a test for no PERL\_NO\_GET\_CONTEXT by itself\.
> But ideally in the future XS&#8203;::APItest would be the only no
> PERL\_NO\_GET\_CONTEXT code included in perl\.

Many of the XS modules supplied with perl do use PERL\_NO\_GET\_CONTEXT\, so it's well tested there\.

Also\, XSUB\.h only does something special when PERL\_NO\_GET\_CONTEXT is \*not\* defined\, so default behaviour without XSUB\.h is otherwise very well tested\, since perl's core code uses the perl API without it\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon May 19 03&#8203;:46&#8203;:12 2014\, jhi wrote&#8203;:
> On Sunday\-201405\-18\, 22&#8203;:08\, Tony Cook via RT wrote&#8203;:
> > a\) your original patch that removes a redundant dTHX
> >
> > The first has the advantage that we continue to have a core
> > maintained module which low performance requirements that will
> > probably break if we somehow manage to break the aTHX re\-definition
> > in XSUB\.h\.
> 
> I think less complex is much more preferable; I don't think speeding
> up
> XS\-APItest is a worthwhile goal\.

Applied by Jarkko as a1b1d0558e4f9a299f56def4b6400355a161005f\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
